exercise-pollitos-shcool-miguel exercise where we implement everything we have learned during the course

Mgl07 opened the pull request:
exercise carried out with the topics seen so far during the course.

chicks-school 

Implementation of the topics seen so far, the exercise covers topics such as creating and configuring beans, classes, controllers and services. The exercise is designed for the instantiation of two schools of the same type, where operations such as creation of students, courses and grades are through http requests, for this a tool such as postman or curl will be used.

In order to test the methods of the application, postman will be used, this tool provides the environment to carry out the tests more comfortably.

An example will be provided for creating a student:

Method URL:
http://localhost:8080/app/school/GerardoInstitute/register/student

request body:
{
  "id": "1",
  "firstName": "Miguel",
  "lastName": "Garcia",
  "age": 34
}

![image](https://github.com/user-attachments/assets/0e385926-f6e8-4fb5-9469-7fa9938a3cbc)

The result should indicate a 200 for the request

![image](https://github.com/user-attachments/assets/951575ff-e683-42a7-a435-13fa51954df6)

Mgl07 committed:
Demo commit and push

Mgl07 committed:
exercise pollitos-school
functional methods
 -create student
 -get students
 -create courses
 -get courses

Mgl07 committed:
exercise pollitos-school
functional methods
 -create student
 -get students
 -create courses
 -get courses

 The other methods requested in the exercise were completed and the use of logs was implemented.

Mgl07 committed:
exercise pollitos-school
functional methods
 -create student
 -get students
 -create courses
 -get courses

 The other methods requested in the exercise were completed and the use of logs was implemented.

The changes discussed were applied during the course of Monday

RogerPerva commented:
Hello, here I wrote some points to evaluate:

1.- The http://localhost:8080/app/GerardoInstitute/grade/student/1 request for GerardoInstitute returns an empty array.
2.- Not sure what do you want to return if a new model of School o the Initialized School object in: 

` @Profile("default")
    @Bean
    @Qualifier("GerardoInstitute")
    public School gerardoInstitute(){
        School school = new School();
        school.setName("GerardoInstitute");    
        return new School();
    }

    @Profile("default")
    @Bean
    @Qualifier("ZetCollege")
    public School zetCollege(){
        School school = new School();
        school.setName("ZetCollege");
        return new School();
    }`


3.- in the schoolController you can implement the responseEntities
4.-in your Service is better to compare with equalsIgnoreCase because the validations could fail.
5.- in getGradesByStudent you can add a message because youre returning a List<String> instead of an empty list
6.- you have a method called createrStudents, not sure if you wanted to named that method as 'creater' or is a finger mistake, but could make issues in case you want to type 'create' or 'created'
7.- youre duplicating code between ZetCollege and GerardoInstitute, you can separete methods in order to reuse code.

thank you

RogerPerva reviewed (commented):
Hello, here I wrote some points to evaluate:

1.- The http://localhost:8080/app/GerardoInstitute/grade/student/1 request for GerardoInstitute returns an empty array.
2.- Not sure what do you want to return if a new model of School o the Initialized School object in:

` @Profile("default")
@bean
@qualifier("GerardoInstitute")
public School gerardoInstitute(){
School school = new School();
school.setName("GerardoInstitute");
return new School();
}

@Profile("default")
@Bean
@Qualifier("ZetCollege")
public School zetCollege(){
    School school = new School();
    school.setName("ZetCollege");
    return new School();
}`
3.- in the schoolController you can implement the responseEntities
4.-in your Service is better to compare with equalsIgnoreCase because the validations could fail.
5.- in getGradesByStudent you can add a message because youre returning a List instead of an empty list
6.- you have a method called createrStudents, not sure if you wanted to named that method as 'creater' or is a finger mistake, but could make issues in case you want to type 'create' or 'created'
7.- youre duplicating code between ZetCollege and GerardoInstitute, you can separete methods in order to reuse code.

thank you

frankisof commented on src/main/java/com/iwaconsolti/school/miguel/config/SchoolConfig.java:
This class is too long and could be confusing, I recommend separating the logic of creating beans and the logic in which you fill them to avoid long classes

frankisof commented on src/main/java/com/iwaconsolti/school/miguel/controller/SchoolController.java:
The controller class must follow the principle of single responsibility, in this class you implement all the endpoints of your program, this causes long and confusing classes. I recommend creating a controller for each DTO implemented

frankisof commented on src/main/java/com/iwaconsolti/school/miguel/service/SchoolService.java:
The service class is too long, it should follow the principles of single responsibility and low coupling, I suggest creating a service for each implemented model, developing a model, controller and service for each DTO simplifies the understanding of the code and avoids long classes

Mgl07 replied:
reorganize the classes for the implementation of the DTO classes

frankisof commented on src/test/resources/pollitos_school.postman_collection.json:
a single postman file is difficult to understand without the requests clearly stated, I recommend putting the requests in the description of your PR or individually exporting each http request in a single file

example:
getcourses.postman_collection
setcourses.postman_collection
ETC

Mgl07 replied:
reorganize postmas collections to make them more readable

frankisof reviewed (commented):
Good morning, I have reviewed your PR which works well, however there are areas that could be better


1.- implement an MVC structure 
2.- implement the principles of single responsibility and low coupling 
3.- avoid the use of very long classes 
4.- implement the principle of separation of responsibilities 
5.- implement a postman collection for each endpoint
6.- implement a data, controller and service for each object

Mgl07 commented:
@RogerPerva I don't understand your observations 2, 5, 7, could you give me more details about what you mean?

Observation 1 gives you that result because the method as such does not exist and was not required at the time.

Mgl07 committed:
exercise pollitos-school
functional methods
 -create student
 -get students
 -create courses
 -get courses

 The other methods requested in the exercise were completed and the use of logs was implemented.

changes that are considered based on the observations made

The changes discussed were applied during the course of Monday

RogerPerva commented on src/main/java/com/iwaconsolti/school/miguel/config/SchoolConfig.java:
```diff
@@ -0,0 +1,33 @@
+    public School gerardoInstitute(){
+        School school = new School();
+        school.setName("GerardoInstitute");
+        return new School();
```
in here, you are setting the name as "Gerardoinstitute" but you are not using it because, i believe you wanted to return the school object with the name already setted as "gerardoinstitute".
```suggestion
        return school();
```

Mgl07 replied:
I already confirmed what was mentioned and it was correct since a new object was returned

Mgl07 replied:
I already confirmed what was mentioned and it was correct since a new object was returned

RogerPerva commented on src/main/java/com/iwaconsolti/school/miguel/service/GradeService.java:
```diff
@@ -0,0 +1,110 @@
+                }
+            }
+        }
+        return studentGrades;
```
5.- In here you can return a Message in case there are no grades

Mgl07 replied:
I understand what you told me but I feel that this is fine since the view is the one that will be in charge of giving the message when the empty list arrives, since that's how I set it up, also if I wanted to return a message I would have to change the type of the method

Mgl07 committed:
exercise pollitos-school
functional methods
 -create student
 -get students
 -create courses
 -get courses

 The other methods requested in the exercise were completed and the use of logs was implemented.

changes that are considered based on the observations made

The changes discussed were applied during the course of Monday

Mgl07 committed:
exercise pollitos-school
functional methods
 -create student
 -get students
 -create courses
 -get courses

 The other methods requested in the exercise were completed and the use of logs was implemented.

changes that are considered based on the observations made

The changes discussed were applied during the course of Monday

Mgl07 committed:
exercise pollitos-school
functional methods
 -create student
 -get students
 -create courses
 -get courses

 The other methods requested in the exercise were completed and the use of logs was implemented.

changes that are considered based on the observations made

The changes discussed were applied during the course of Monday

Mgl07 committed:
exercise pollitos-school
functional methods
 -create student
 -get students
 -create courses
 -get courses

 The other methods requested in the exercise were completed and the use of logs was implemented.

changes that are considered based on the observations made

The changes discussed were applied during the course of Monday

fjpgtt commented on src/main/java/com/iwaconsolti/school/miguel/config/SchoolConfig.java:
```diff
@@ -0,0 +1,33 @@
+@Slf4j
+public class SchoolConfig {
+
+    @Profile("default")
```
Why you dont put the Profile at class level?

fjpgtt commented on src/main/java/com/iwaconsolti/school/miguel/config/SchoolPopulatedConfig.java:
```diff
@@ -0,0 +1,97 @@
+    @Bean
+    @Qualifier("GerardoInstitute")
+    public School gerardoInstitutePopulated(){
+        log.info("The populated profile is activated, the data will be loaded");
```
On of these 2 logs should be DEBUG or deleted, since you can see that is executed by the other log that you keep

fjpgtt commented on src/main/java/com/iwaconsolti/school/miguel/model/School.java:
```diff
@@ -0,0 +1,13 @@
+
+@Data
+public class School {
+    private String name;
```
You could make this field private so the constructor require always to pass the name, since is not possible to create the school without a name

fjpgtt commented on src/main/java/com/iwaconsolti/school/miguel/model/Students.java:
```diff
@@ -0,0 +1,10 @@
+
+@Data
+public class Students {
+    private int id;
```
From what i see you always pass all the values when you create the Students, so maybe is better to have AllArgsConstructor to avoid having to use setters always

fjpgtt commented on src/main/java/com/iwaconsolti/school/miguel/model/dto/CoursesDTO.java:
```diff
@@ -0,0 +1,9 @@
+import lombok.Data;
+
+@Data
+public class CoursesDTO {
```
You should have a constructor that receive the Course object and popualte the values, that way you dont have to use Setters

fjpgtt commented on src/main/java/com/iwaconsolti/school/miguel/config/SchoolPopulatedConfig.java:
```diff
@@ -0,0 +1,97 @@
+        school.getCourses().put(firstCourse.getId(), firstCourse);
+        school.getCourses().put(secondCourse.getId(), secondCourse);
+
+        log.info("I finished loading the data");
```
This log can be removed or reduced to debug

fjpgtt commented on src/main/java/com/iwaconsolti/school/miguel/controller/CourseController.java:
```diff
@@ -0,0 +1,53 @@
+    public ResponseEntity<Courses> newCourse(@PathVariable String nameSchool, @RequestBody CoursesDTO courseDTO) {
+
+        if (courseDTO.getNameCourse() == null || courseDTO.getProfessorName() == null){
+            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(null);
```
In the body you can return a message so the users know what error happen

fjpgtt commented on src/main/java/com/iwaconsolti/school/miguel/service/CourseService.java:
```diff
@@ -0,0 +1,85 @@
+        course.setProfessorName(courseDTO.getProfessorName());
+
+        if("GerardoInstitute".equalsIgnoreCase(schoolName)){
+            gerardoInstitute.getCourses().put(course.getId(), course);
```
If I pass an id that already in the map, it will overwrite the previous one. I think you can validate that it doesnt exists before doing the put

fjpgtt commented on src/main/java/com/iwaconsolti/school/miguel/service/CourseService.java:
```diff
@@ -0,0 +1,85 @@
+
+    public Courses createCourse(String schoolName, CoursesDTO courseDTO) {
+
+        Courses course = new Courses();
```
You should have a constructor that receive a DTO and create the course, to avoid duplicate code

fjpgtt commented on src/main/java/com/iwaconsolti/school/miguel/service/CourseService.java:
```diff
@@ -0,0 +1,85 @@
+
+        if("GerardoInstitute".equalsIgnoreCase(schoolName)){
+            for(Courses courses : gerardoInstitute.getCourses().values()){
+                CoursesDTO courseDTO = new CoursesDTO();
```
You should have a method or a constructor that receive the Course and return the DTO populated

fjpgtt commented on src/main/java/com/iwaconsolti/school/miguel/service/CourseService.java:
```diff
@@ -0,0 +1,85 @@
+
+    public Courses editCourse(int id, String schoolName, CoursesDTO courseDTO){
+
+        Courses course = new Courses();
```
Instead of creating a new object you should get the element from the map, and then do the set if it exists. right now I can sent an entry with an id that doesnt exists and is added to the map

fjpgtt commented on src/main/java/com/iwaconsolti/school/miguel/controller/CourseController.java:
```diff
@@ -0,0 +1,53 @@
+        Collection<CoursesDTO> courses = courseService.getCourses(nameSchool);
+
+        if(courses.isEmpty()){
+            return ResponseEntity.status(HttpStatus.NOT_FOUND).body(null);
```
In this case an empty courses is a correct value, if you havent added any course yet

fjpgtt commented on src/main/java/com/iwaconsolti/school/miguel/controller/CourseController.java:
```diff
@@ -0,0 +1,53 @@
+
+        Courses updateCourse = courseService.editCourse(courseId, nameSchool, courseDTO);
+
+        if(updateCourse == null){
```
Right now this is not possible since you always return an object from the editCourse

fjpgtt reviewed (changes requested):
NIT: the package path should not include the miguel part

You need to add a better validation to check if the name of the school exists

fjpgtt commented on src/main/java/com/iwaconsolti/school/miguel/service/StudentService.java:
```diff
@@ -0,0 +1,114 @@
+        student.setAge(studentDTO.getAge());
+
+        if("GerardoInstitute".equalsIgnoreCase(schoolName)){
+            if(gerardoInstitute.getStudents().containsKey(student.getId())){
```
Why is this only in gerardoInstitute and not in the zetCollege

fjpgtt commented on src/main/java/com/iwaconsolti/school/miguel/service/StudentService.java:
```diff
@@ -0,0 +1,114 @@
+
+    public Collection<StudentsDTO> getStudentById(String schoolName, int studentId){
+
+        Collection<StudentsDTO> studentsDTOS = new ArrayList<>();
```
If you pass a studentId, you should only return one student

fjpgtt commented on src/main/java/com/iwaconsolti/school/miguel/controller/StudentController.java:
```diff
@@ -0,0 +1,75 @@
+import java.util.Map;
+
+@RestController
+@RequestMapping("/app/{nameSchool}")
```
student should be part of the class requestmapping

fjpgtt commented on src/main/java/com/iwaconsolti/school/miguel/controller/CourseController.java:
```diff
@@ -0,0 +1,53 @@
+import java.util.Collection;
+
+@RestController
+@RequestMapping("/app/{nameSchool}")
```
course should be part of this requestmapping

fjpgtt commented on src/main/java/com/iwaconsolti/school/miguel/controller/StudentController.java:
```diff
@@ -0,0 +1,75 @@
+
+        if(studentDTO.getFirstName() == null || studentDTO.getLastName() == null || studentDTO.getAge() == 0){
+            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(
+                    Map.of("message","Required data is missing","status",HttpStatus.BAD_REQUEST.value())
```
status is not required since that is in the response status

fjpgtt commented on src/main/java/com/iwaconsolti/school/miguel/controller/StudentController.java:
```diff
@@ -0,0 +1,75 @@
+            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(
+                    Map.of("message","Required data is missing","status",HttpStatus.BAD_REQUEST.value())
+            );
+        } else if(studentService.getStudentById(nameSchool,studentDTO.getId()) != null){
```
This can be a simple if, because if the previous if enter, the endpoint ends, because it has a return

fjpgtt commented on src/main/java/com/iwaconsolti/school/miguel/controller/StudentController.java:
```diff
@@ -0,0 +1,75 @@
+    @GetMapping("/student/{studentId}")
+    public ResponseEntity<Collection<StudentsDTO>> getStudentById(@PathVariable String nameSchool,@PathVariable int studentId){
+
+        Collection<StudentsDTO> getStudent = studentService.getStudentById(nameSchool,studentId);
```
Here you only want one value returned since the id should not be duplicated

fjpgtt commented on src/main/java/com/iwaconsolti/school/miguel/controller/GradeController.java:
```diff
@@ -0,0 +1,59 @@
+    @PostMapping("/grade")
+    public ResponseEntity<Grade> createGrade(@PathVariable String nameSchool, @RequestBody GradesDTO gradeDTO){
+
+        if(gradeDTO.getScore() < 0 || gradeDTO.getScore() > 100){
```
This 100 should be the property that you passed

fjpgtt commented on src/main/java/com/iwaconsolti/school/miguel/service/GradeService.java:
```diff
@@ -0,0 +1,143 @@
+        if("GerardoInstitute".equalsIgnoreCase(schoolName)){
+            for(Grade grade : gerardoInstitute.getGrades().values()) {
+                if (grade.getStudent().getId() == id) {
+                    studentGrades.add("Name: " + grade.getStudent().getFirstName() + " " + grade.getStudent().getLastName());
```
Instead of a list of string is better to create an object with the values so it is in a better json structure

fjpgtt commented on src/main/java/com/iwaconsolti/school/miguel/service/GradeService.java:
```diff
@@ -0,0 +1,143 @@
+        }else if("ZetCollege".equalsIgnoreCase(schoolName)) {
+            for(Grade grade: zetCollege.getGrades().values()) {
+                if (grade.getStudent().getId() == id) {
+                    studentGrades.add("Name: " + grade.getStudent().getFirstName() + " " + grade.getStudent().getLastName());
```
This and the code above is duplicated, it could have being extracted to a private method

fjpgtt commented on src/main/java/com/iwaconsolti/school/miguel/service/GradeService.java:
```diff
@@ -0,0 +1,143 @@
+        boolean removed = false;
+
+        if("GerardoInstitute".equalsIgnoreCase(schoolName)){
+            for(int key : new ArrayList<>(gerardoInstitute.getGrades().keySet())){
```
Any reason why you did use the set returned and you convert it to a list?

fjpgtt replied:
```suggestion
                        for(Map.Entry<Integer, Grade> entry : gerardoInstitute.getGrades().entrySet()){
                if(entry.getValue().getCourse().getId() == courseId){
                    gerardoInstitute.getGrades().remove(entry.getKey());
```

Mgl07 committed:
implementation of the repositories to the pollitos-school exercise
the recommendations/changes made in the pull request were applied
